sfepy/discrete/fem/fields_base.py

```diff
@@ -1438,42 +1257,53 @@ def average_qp_to_vertices(self, data_qp, integral):
         """
         region = self.region
 
-        n_cells = region.get_n_cells(True)
+        n_cells = region.get_n_cells(is_surface=self.is_surface)
         if n_cells != data_qp.shape[0]:
             msg = 'incomatible shape! (%d == %d)' % (n_cells,
                                                      data_qp.shape[0])
             raise ValueError(msg)
 
-        n_vertex = len(region.vertices)
+        n_vertex = self.n_vertex_dof
         nc = data_qp.shape[2]
 
         nod_vol = nm.zeros((n_vertex,), dtype=nm.float64)
         data_vertex = nm.zeros((n_vertex, nc), dtype=nm.float64)
 
-        sg = self.get_mapping(self.region, integral, 'surface')[0]
+        rg = self.get_mapping(self.region, integral, region.kind)[0]
 
-        area = nm.squeeze(sg.volume)
-        n_cells = region.get_n_cells(True)
-        iels = nm.arange(n_cells, dtype=nm.int32)
+        area = nm.squeeze(rg.volume)
+        iels = self.region.get_cells()
 
         data_e = nm.zeros((area.shape[0], 1, nc, 1), dtype=nm.float64)
-        sg.integrate(data_e, data_qp[iels])
+        rg.integrate(data_e, data_qp[iels])
 
         ir = nm.arange(nc, dtype=nm.int32)
 
-        sd = self.domain.surface_groups[region.name]
-        # Should be vertex connectivity!
-        conn = sd.get_connectivity(local=True)
+        if region.kind == 'cell':
+            conn = self.econn[:, :self.gel.n_vertex]
+        elif region.kind == 'facet':
+            sd = self.domain.surface_groups[region.name]
+            # Should be vertex connectivity!
+            conn = sd.get_connectivity(local=True)
+
         for ii, cc in enumerate(conn):
             # Assumes unique nodes in cc!
             ind2, ind1 = nm.meshgrid(ir, cc)
-            data_vertex[ind1,ind2] += data_e[iels[ii],0,:,0]
+            data_vertex[ind1, ind2] += data_e[iels[ii], 0, :, 0]
             nod_vol[cc] += area[ii]
 
-        data_vertex /= nod_vol[:,nm.newaxis]
+        data_vertex /= nod_vol[:, nm.newaxis]
 
         return data_vertex
 
+    # from SurfaceField()
+    # def _setup_bubble_dofs(self):
+    #     """
+    #     Setup bubble DOF connectivity.
+    #     """
+    #     return 0, None, None
+
```

Comment:
SurfaceField was defined on faces of a volume mesh, so it did not have bubble DOFs, only vertex, edge and face. How is this solved now?

Reply:
What should return this function for `VolumeField` (`self.is_surface` is `False`)?

Reply:
See e.g. https://github.com/sfepy/sfepy/blob/master/sfepy/discrete/fem/fields_nodal.py